### PR TITLE
Use memchr for str::find(char)

### LINKED ISSUE #46693: str::find(char) is slower than it ought ot be
If you try to find a specific character in a string, it performs significantly worse than the theoretical optimal implementation built on `memchr`.

For example, consider the following benchmark. On my laptop the `find` method runs at about 90ns/iter, and the `memchr` method runs at 4ns/iter. I would imagine that we should optimize them such that they are the same.

```rust
#![feature(test)]
extern crate memchr;
extern crate test;

use test::{black_box, Bencher};

const DEMO_STRING: &str = "this is a string with a decent number of ascii characters and \n there is a new line in the middle which it should find";

#[bench]
fn bench_find(b: &mut Bencher) {
    b.iter(|| {
        let s = test::black_box(DEMO_STRING);
        s.find('\n')
    });
}

#[bench]
fn bench_memchr(b: &mut Bencher) {
    b.iter(|| {
        let s = test::black_box(DEMO_STRING);
        memchr::memchr(b'\n', s.as_bytes())
    });
}
```

### PR DESCRIPTION
**Pull request by Manishearth**
This is a 10x improvement for searching for characters.

This also contains the patches from https://github.com/rust-lang/rust/pull/46713 . Feel free to land both separately or together.

cc @mystor @alexcrichton

r? @bluss

fixes #46693

### CODE REVIEW

**Comment by Manishearth** [comment]
I haven't really tested this much, there probably are failures. Will do a second pass at self-review once I know we pass all tests (from travis)

**Comment by Manishearth** [comment]
The `memchr` crate is even faster because it links to glibc's memchr, which uses SIMD and other fancy stuff. libcore can't link to this so to get these wins we'll have to do a SIMD impl ourselves.

```rust
#[bench]
fn find_char(b: &mut Bencher) {
    let x = test::black_box("Lorem Ipsum is simply dummy text of the printing and typesetting industry. Lorem Ipsum has been the industry's standard dummy text ever since the 1500s, when an unknown printer took a galley of type and scrambled it to make a type specimen book. It has survived not only five centuries, but also the leap into electronic typesetting, remaining essentially unchanged. It was popularised in the 1960s with the release of Letraset sheets containing Lorem Ipsum passages, and more recently with desktop publishing software like Aldus PageMaker including versions of Lorem Ipsum.");
    b.iter(|| test::black_box(x.find('/')));
}


#[bench]
fn find_char_memchr(b: &mut Bencher) {
    let x = test::black_box("Lorem Ipsum is simply dummy text of the printing and typesetting industry. Lorem Ipsum has been the industry's standard dummy text ever since the 1500s, when an unknown printer took a galley of type and scrambled it to make a type specimen book. It has survived not only five centuries, but also the leap into electronic typesetting, remaining essentially unchanged. It was popularised in the 1960s with the release of Letraset sheets containing Lorem Ipsum passages, and more recently with desktop publishing software like Aldus PageMaker including versions of Lorem Ipsum.");
    b.iter(|| test::black_box(memchr::memchr(b'/', x.as_bytes())));
}
```


Before:

```
running 2 tests
test find_char        ... bench:         593 ns/iter (+/- 201)
test find_char_memchr ... bench:           9 ns/iter (+/- 1)
```

After:
```
running 2 tests
test find_char        ... bench:          57 ns/iter (+/- 12)
test find_char_memchr ... bench:           9 ns/iter (+/- 1)
```

**Comment by Manishearth** [comment]
This does not bring improvements for multibyte chars or for `str::find(str)`. We _can_ bring improvement for these, but it's tricky.

For `str` when it starts with an ASCII char we can do similar stuff as here (and then use the original algorithm to finish the match.

When the thing we're searching for is _not_ ASCII we can still search for the first byte. However for most UTF8 text the first byte will generally be pretty uniform; i.e. if it's Arabic text will usually be 0xD8 or 0xD9, Korean will be 0xEA, 0xEB, 0xEC, or 0xED, Devanagari is usually 0xE0, etc. This means that memchr will have lots of false positives; we'll get lots of hits on the first byte and then have to check the second byte. This amount of stutter will probably make memchr's (minor) fixed overhead significant, and destroy any perf gains which we may get.

Searching for the second byte or even better, the _last_ byte, might work better. But I'm not sure if I want to write that code right now, and the tradeoffs are a bit trickier there :)

**Comment by Manishearth** [comment]
Bench numbers do not materially change with the UTF8 changes. I did come up with a pathological case of searching a Devanagari string for `ä` (which shares bytes) that ends up being 2x slower because every other character is a false positive hit (entirely negating memchr's win).

I think this pathological case is ok, it will only arise when mixing languages and for very specific characters.

I can check some form of these benchmarks into tree if y'all feel it necessary.

```
$ cargo bench
test find_char                            ... bench:         603 ns/iter (+/- 203)
test find_char_memchr                     ... bench:          10 ns/iter (+/- 3)
test find_multibyte_char_found            ... bench:         376 ns/iter (+/- 67)
test find_multibyte_char_notfound         ... bench:         618 ns/iter (+/- 129)
test find_multibyte_string_multibyte_char ... bench:         719 ns/iter (+/- 137)
test find_multibyte_string_pathological   ... bench:         620 ns/iter (+/- 98)

$ cargo +x-stage2 bench
test find_char                            ... bench:          67 ns/iter (+/- 45)
test find_char_memchr                     ... bench:          10 ns/iter (+/- 1)
test find_multibyte_char_found            ... bench:          50 ns/iter (+/- 12)
test find_multibyte_char_notfound         ... bench:          74 ns/iter (+/- 20)
test find_multibyte_string_multibyte_char ... bench:          74 ns/iter (+/- 20)
test find_multibyte_string_pathological   ... bench:       1,672 ns/iter (+/- 348)
```

Code:
<details>
```rust

#[bench]
fn find_char(b: &mut Bencher) {
    let x = test::black_box("Lorem Ipsum is simply dummy text of the printing and typesetting industry. Lorem Ipsum has been the industry's standard dummy text ever since the 1500s, when an unknown printer took a galley of type and scrambled it to make a type specimen book. It has survived not only five centuries, but also the leap into electronic typesetting, remaining essentially unchanged. It was popularised in the 1960s with the release of Letraset sheets containing Lorem Ipsum passages, and more recently with desktop publishing software like Aldus PageMaker including versions of Lorem Ipsum.");
    b.iter(|| test::black_box(x.find('/')));
}

#[bench]
fn find_char_memchr(b: &mut Bencher) {
    let x = test::black_box("Lorem Ipsum is simply dummy text of the printing and typesetting industry. Lorem Ipsum has been the industry's standard dummy text ever since the 1500s, when an unknown printer took a galley of type and scrambled it to make a type specimen book. It has survived not only five centuries, but also the leap into electronic typesetting, remaining essentially unchanged. It was popularised in the 1960s with the release of Letraset sheets containing Lorem Ipsum passages, and more recently with desktop publishing software like Aldus PageMaker including versions of Lorem Ipsum.");
    b.iter(|| test::black_box(memchr::memchr(b'/', x.as_bytes())));
}

#[bench]
fn find_multibyte_char_found(b: &mut Bencher) {
    let x = test::black_box("Lorem Ipsum is simply dummy text of the printing and typesetting industry. Lorem Ipsum has been the industry's standard dummy text ever since the 1500s, when an unknown printer took a galley of type and scrambled it to make a type specimen book. It has survived not only five centuries, but also the leap into electronic typesetting, ก remaining essentially unchanged. It was popularised in the 1960s with the release of Letraset sheets containing Lorem Ipsum passages, and more recently with desktop publishing software like Aldus PageMaker including versions of Lorem Ipsum.");
    b.iter(|| test::black_box(x.find('ก')));
}

#[bench]
fn find_multibyte_char_notfound(b: &mut Bencher) {
    let x = test::black_box("Lorem Ipsum is simply dummy text of the printing and typesetting industry. Lorem Ipsum has been the industry's standard dummy text ever since the 1500s, when an unknown printer took a galley of type and scrambled it to make a type specimen book. It has survived not only five centuries, but also the leap into electronic typesetting, remaining essentially unchanged. It was popularised in the 1960s with the release of Letraset sheets containing Lorem Ipsum passages, and more recently with desktop publishing software like Aldus PageMaker including versions of Lorem Ipsum.");
    b.iter(|| test::black_box(x.find('ก')));
}

#[bench]
fn find_multibyte_string_multibyte_char(b: &mut Bencher) {
    let x = test::black_box("जलद कोल्हा आळशी कुत्रा वरुन उडी मारली, जलद कोल्हा आळशी कुत्रा वरुन उडी मारली, जलद कोल्हा आळशी कुत्रा वरुन उडी मारली, जलद कोल्हा आळशी कुत्रा वरुन उडी मारली, जलद कोल्हा आळशी कुत्रा वरुन उडी मारली, जलद कोल्हा आळशी कुत्रा वरुन उडी मारली, जलद कोल्हा आळशी कुत्रा वरुन उडी मारली");
    b.iter(|| test::black_box(x.find('ग'))); // not in the string
}

#[bench]
fn find_multibyte_string_pathological(b: &mut Bencher) {
    let x = test::black_box("जलद कोल्हा आळशी कुत्रा वरुन उडी मारली, जलद कोल्हा आळशी कुत्रा वरुन उडी मारली, जलद कोल्हा आळशी कुत्रा वरुन उडी मारली, जलद कोल्हा आळशी कुत्रा वरुन उडी मारली, जलद कोल्हा आळशी कुत्रा वरुन उडी मारली, जलद कोल्हा आळशी कुत्रा वरुन उडी मारली, जलद कोल्हा आळशी कुत्रा वरुन उडी मारली");
    b.iter(|| test::black_box(x.find('ä'))); // ä's last byte is found often in Devanagari text
}
```
</details>

**Comment by Manishearth** [comment]
If we really care about the pathological case it can be avoided by having some check in the loop that after X false positives falls back to regular "loop on next" behavior

**Comment by Manishearth** [comment]
@rust-lang/infra could I get a perf.rlo diff result from this try push?

**Comment by nagisa** [comment]
cc @Mark-Simulacrum ^

**Comment by BurntSushi** [comment]
> When the thing we're searching for is not ASCII we can still search for the first byte. However for most UTF8 text the first byte will generally be pretty uniform; i.e. if it's Arabic text will usually be 0xD8 or 0xD9, Korean will be 0xEA, 0xEB, 0xEC, or 0xED, Devanagari is usually 0xE0, etc. This means that memchr will have lots of false positives; we'll get lots of hits on the first byte and then have to check the second byte. This amount of stutter will probably make memchr's (minor) fixed overhead significant, and destroy any perf gains which we may get.

> Searching for the second byte or even better, the last byte, might work better. But I'm not sure if I want to write that code right now, and the tradeoffs are a bit trickier there :)

Searching for the last byte is indeed a better heuristic on UTF-8 than searching for the first byte. You'd be in good company (GNU grep does that). But the last byte is still arbitrary. This is why the regex crate ranks every byte in order of what it believes is rare. Leading UTF-8 bytes are considered common while trailing bytes aren't. But you also get things like "`z` is rarer than `a`," which it commonly is. So the `memchr` is applied to the rarest byte in the pattern. Of course, you still wind up with pathological cases when the frequency rank doesn't match the corpus, but this will _always_ be true when using `memchr` without analyzing the haystack before hand (which obviously doesn't make sense in this specific domain of text search).

File: src/libcore/str/pattern.rs [hunk 1]
```diff
@@ -234,70 +243,275 @@ pub unsafe trait ReverseSearcher<'a>: Searcher<'a> {
...
-impl CharEq for char {
+unsafe impl<'a> Searcher<'a> for CharSearcher<'a> {
+    #[inline]
+    fn haystack(&self) -> &'a str {
+        self.haystack
+    }
+    #[inline]
+    fn next(&mut self) -> SearchStep {
+        let old_finger = self.finger;
+        let slice = unsafe { self.haystack.get_unchecked(old_finger..self.haystack.len()) };
```
Comment (BurntSushi) [inline src/libcore/str/pattern.rs]: Do the various bounds check elisions actually help here? I've tried eliding them in my own substring search algorithms and it meets with variable success.

**Comment by BurntSushi** [comment]
This LGTM! Nice work @Manishearth :-)

**Comment by BubbaSheen** [comment]
very helpful indeed gentelman

File: src/libcore/str/pattern.rs (hunk 1 above)
Comment (Manishearth) [reply src/libcore/str/pattern.rs]: I don't think they do, but I haven't checked and it seemed pretty easy to keep that invariant. I can check if you want.

**Comment by Manishearth** [comment]
> You'd be in good company (GNU grep does that). 

yay :)

> To be clear, I think the frequency rank stuff is probably overkill for searching a single char and would probably just stick to the last byte.

_phew_

that sounds trickier to get right 😄 

> Do we really not already use memchr in str::find(str) though?

Yeah, we do an interesting but non-memchry algorithm. I considered retrofitting the existing memchr'd `.find(char)` into `.find(str)` but that would mean losing the existing algorithm which means the wins are iffier (not to mention that memchr has very little wins if you're stuttering the algorithm all the time, which is far likelier with a `.find(str)` built on top of `.find(char)`)


> This LGTM! Nice work @Manishearth :-)

can this be landed r=you? I've made a small mistake which I need to rectify, aside from that it seems basically ready. Or should we wait for second review?

**Comment by BurntSushi** [comment]
@Manishearth Yeah r=me sounds great.

File: src/libcore/str/pattern.rs (hunk 1 above)
Comment (BurntSushi) [reply src/libcore/str/pattern.rs]: My general position has been to not elide bounds checks unless I'm pretty sure that it matters. If it were me, I'd remove the `unsafe`. :)

**Comment by Mark-Simulacrum** [comment]
Perf queued; in the future please ping me directly.

File: src/libcore/str/pattern.rs (hunk 1 above)
Comment (Manishearth) [reply src/libcore/str/pattern.rs]: I'll do some checking later this week, for now I'll land it.

File: src/libcore/slice/memchr.rs [hunk 2]
```diff
@@ -0,0 +1,230 @@
...
+#[cfg(target_pointer_width = "64")]
+#[inline]
+fn repeat_byte(b: u8) -> usize {
+    let mut rep = (b as usize) << 8 | b as usize;
+    rep = rep << 16 | rep;
+    rep = rep << 32 | rep;
+    rep
+}
+
+/// Return the first index matching the byte `a` in `text`.
```
Comment (jesseschalken) [inline src/libcore/slice/memchr.rs]: `a` is meant to be `x`?

Comment (Manishearth) [reply src/libcore/slice/memchr.rs]: yeah, fixing

Comment (Manishearth) [reply src/libcore/slice/memchr.rs]: someone fixed it already

**Pull request merged**